solr/contrib/blob-repository/src/java/org/apache/solr/s3/AdobeMockS3StorageClient.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.s3;
+
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+/**
+ * This storage client exists to work around some of the incongruencies Adobe S3Mock has with the S3 API.
+ * The main difference is that S3Mock does not support paths with a leading '/', but S3 does, and our code
+ * in {@link S3StorageClient} requires all paths to have a leading '/'.
+ */
+class AdobeMockS3StorageClient extends S3StorageClient {
+
+    static final int DEFAULT_MOCK_S3_PORT = 9090;
+    private static final String DEFAULT_MOCK_S3_ENDPOINT = "http://localhost:" + DEFAULT_MOCK_S3_PORT;
+
+    AdobeMockS3StorageClient(String bucketName) {
+        super(createInternalClient(), bucketName);
+    }
+
+    @VisibleForTesting
+    AdobeMockS3StorageClient(AmazonS3 s3client, String bucketName) {
+        super(s3client, bucketName);
+    }
+
+    private static AmazonS3 createInternalClient() {
+        String s3MockEndpoint = System.getenv().getOrDefault("MOCK_S3_ENDPOINT", DEFAULT_MOCK_S3_ENDPOINT);
+
+        return AmazonS3ClientBuilder.standard()
+            .enablePathStyleAccess()
+            .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(s3MockEndpoint, Regions.US_EAST_1.name()))
+            .build();
+    }
+
+    /**
+     * Ensures path adheres to some rules (different than the rules that S3 cares about):
+     * -Trims leading slash, if given
+     * -If it's a file, throw an error if it ends with a trailing slash
+     */
+    @Override
+    String sanitizedPath(String path, boolean isFile) throws S3Exception {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(path));
```

Comment:
> pulling in Guava just for Preconditions

Do you mean in this particular class? Or in general? I would definitely agree if we pulled Guava newly into Solr for just a few convenience functions, but as-is, it's already there on the classpath (same as Apache StringUtils)

Reply:
This line should also not be here, it's causing the example in `solr/contrib/blob-repository/README.md` to fail, because the `location` path gets sanitized to an empty string, which fails. The failures should occur when someone tries to write to an empty path, such as when `location` and `name` are empty, but that can be caught much earlier on, such as in the collections API request handler. Or when the S3MockClient actually tries to make the request....

Reply:
> it's causing the example in solr/contrib/blob-repository/README.md to fail

Interesting, I tested the example before pushing and it is working for me as written.